Support sp_rename for OBJECT type

raydhkim opened the pull request:
Previously, Babelfish did not support sp_rename for Babelfish objects, which prevented the users from renaming objects and table columns. It is important to support sp_rename to allow users change and modify their own-created objects in the database. Now, sp_rename for OBJECT type is supported, allowing users to rename their own-created TABLE, VIEW, PROCEDURE, FUNCTION.

Task: BABEL-3888

Signed-off-by: Ray Kim <raydhkim@amazon.com>

### Description

[Describe what this change achieves - Guidelines below (please delete the guidelines after writing the PR description)]

> 1. *What* is the change? This is best described in terms of “Currently, Babelfish does X. With this change it now does Y.” Think of “What *did* it *used* to do?” and “What *does* it do *now*?”
2. *Why* was the change made? What drove our desire to put effort into the change?
3. *How* was the code changed should only appear for large commits. This can serve as a rough roadmap to what’s contained in the commit. It should be very high level; if it’s directly referencing code it’s probably too detailed. It’s also critical that this section of a commit message does not try to replace proper code documentation (ie, block comments or README files). Generally, this section should only appear if the commit itself is large enough that it’s helpful to provide a roadmap to someone looking at the commit.
4. The last descriptive piece is the “title” for the commit: the very first line of the commit message, which should typically be less than 80 characters. A good title is *critical*, because it’s the only thing that shows up in places like the Github commit listing. No one’s got time to read through full commit messages when trying to find a single commit out of dozens.


### Issues Resolved

[List any issues this PR will resolve]

### Test Scenarios Covered ###
* **Use case based -**


* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

macwoody commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2891,3 +2891,60 @@ END;
+		END
+    ELSE
+		BEGIN
+			SET @schemaname = 'dbo';
```
this is wrong, it should be current default schema

macwoody commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2891,3 +2891,60 @@ END;
+		END
+	SELECT @currtype = type FROM sys.objects o1 INNER JOIN sys.schemas s1 ON o1.schema_id = s1.schema_id 
+	WHERE s1.name = @schemaname AND o1.name = @subname;
+	PRINT 'currtype is: ';
```
should not have print in production code

macwoody commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2891,3 +2891,60 @@ END;
+CREATE OR REPLACE PROCEDURE sys.sp_rename(
+    IN "@objname" sys.nvarchar(776),
+    IN "@newname" sys.SYSNAME,
+    IN "@objtype" sys.varchar(13) DEFAULT NULL
```
Need a feature not supported error if objectype is not supported at the moment

macwoody commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2308,3 +2311,236 @@ get_db_owner_role_name(const char *dbname)
+	usertuple = heap_getnext(tblscan, ForwardScanDirection);
+
+	if (!HeapTupleIsValid(usertuple))
+		ereport(ERROR,
```
before throwing exception, need to close opened table / scan

macwoody commented on contrib/babelfishpg_tsql/src/procedures.c:
```diff
@@ -2062,3 +2067,180 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
+		CommandCounterIncrement();
+
+		// parsetree_list = gen_sp_rename_subcmds(obj_name, new_name, schema_name, objtype);
+		parsetree_list = gen_sp_rename_subcmds(obj_name, new_name, schema_name, objtype_code);
```
I think you should only generate one subcmd

xhfanhe commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3163,18 +3163,48 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
+		{
+			if (sql_dialect == SQL_DIALECT_TSQL)
 			{
-				if (prev_ProcessUtility)
-					prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
-							queryEnv, dest, qc);
-				else
-					standard_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
-							queryEnv, dest, qc);
```
We should run this in PSQL dialect.

xhfanhe commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3163,18 +3163,48 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
+				Oid			dbo_id;
+				Oid prev_current_user;
 
-				check_extra_schema_restrictions(parsetree);
```
Same as above, should keep it for PSQL dialect

xhfanhe commented on contrib/babelfishpg_tsql/src/procedures.c:
```diff
@@ -1597,6 +1599,10 @@ gen_sp_addrole_subcmds(const char *user)
 	Node *stmt;
 	CreateRoleStmt *rolestmt;
 	List *user_options = NIL;
+	// NodeTag     type;
+    //  RoleStmtType stmt_type;     /* ROLE/USER/GROUP */
+    //  char       *role;           /* role name */
+    //  List       *options;        /* List of DefElem nodes */
```
Please cleanup these code

xhfanhe commented on contrib/babelfishpg_tsql/src/procedures.c:
```diff
@@ -2062,3 +2067,180 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
+	} else if (objtype == OBJECT_PROCEDURE) {
+		appendStringInfo(&query, "ALTER PROCEDURE dummy RENAME TO dummy; ");
+	// } else if (objtype == OBJECT_FUNCTION) {
+	} else {
```
Need to throw an error if subtype is not supported

xhfanhe commented on test/JDBC/input/storedProcedures/Test-sp_rename-vu-prepare.sql:
```diff
@@ -0,0 +1,58 @@
+-- tsql
```
Please put this file in all upgrade schedules.

We also need dependency tests, you may name the new files as `Test-sp_rename-dep-vu-*`. The dependency prepare script should have a procedure, a function and a view that depend on sp_rename. Note that the dependency test needs to be added to latest schedule.
See https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/610/files#diff-272b1045c86970661de855a946fe1f1eb1b40db189b48e61239cc2805189855e as an example

xhfanhe commented on test/JDBC/input/storedProcedures/Test-sp_rename-vu-verify.sql:
```diff
@@ -0,0 +1,114 @@
+
+EXEC sp_rename 'sp_rename_schema1.sp_rename_view1', 'sp_rename_view1_new', 'OBJECT';
+GO
+
```
Please add a test where the last argument is not specified

xhfanhe commented on test/JDBC/input/storedProcedures/Test-sp_rename-vu-verify.sql:
```diff
@@ -0,0 +1,114 @@
+SELECT * FROM information_schema.tables
+GO
+
+SELECT schema_name, object_name FROM sys.babelfish_view_def
```
Please use ORDER BY clause for SELECT statements, otherwise the output can be in different order. 
And only SELECT from entries that are related to your test by adding restriction, this is to avoid impact from other test scripts. Something like
```
WHERE object_name LIKE '%sp_rename%'
```
Please check all the SELECT in your test

raydhkim committed:
Support sp_rename for OBJECT

Previously, Babelfish did not support sp_rename for Babelfish objects,
which prevented the users from renaming objects and table columns. It is
important to support sp_rename to allow users change and modify their
own-created objects in the database. Now, sp_rename for OBJECT type is
supported, allowing users to rename their own-created TABLE, VIEW,
PROCEDURE, FUNCTION.

Task: BABEL-3888

Signed-off-by: Ray Kim <raydhkim@amazon.com>

xhfanhe commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3012,18 +3012,56 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
+				
+				check_extra_schema_restrictions(parsetree);
+			}
+			break;
```
We should use return here instead of break

macwoody commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2903,3 +2903,93 @@ END;
 LANGUAGE 'pltsql';
 GRANT ALL ON PROCEDURE sys.sp_sproc_columns_100 TO PUBLIC;
+
+CREATE OR REPLACE PROCEDURE sys.sp_rename_internal(
```
babelfish_sp_rename_internal

macwoody commented on contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql:
```diff
@@ -2903,3 +2903,93 @@ END;
+LANGUAGE 'pltsql'
+AS $$
+BEGIN
+	If @objtype IS NULL
```
in the future, once you are familiar with this, entire pltsql part could be implemented with C, which won't be significantly more difficult, but will be much more efficient.

macwoody commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql:
```diff
@@ -1513,4 +1602,4 @@ LANGUAGE C STABLE;
 
 -- Reset search_path to not affect any subsequent scripts
-SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
+SELECT set_config('search_path', trim(leading 'sys, ' from current_setting('search_path')), false);
```
what is diff for this line?

macwoody replied:
oh, you added one space to the end.

raydhkim replied:
I don't see any space and I'm not sure why it recognizes as a change still..

macwoody commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2308,3 +2312,278 @@ get_db_owner_role_name(const char *dbname)
+	bbf_view_def_dsc = RelationGetDescr(bbf_view_def_rel);
+
+	// serach for the row for update => build the key
+	ScanKeyInit(&key[0],
```
I feel this is not correct. 
sys.babelfish_view_def's primary key is dbid, schema_name, object_name, 
plz refer to this search_bbf_view_def function

macwoody commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2308,3 +2312,278 @@ get_db_owner_role_name(const char *dbname)
+				BTEqualStrategyNumber, F_TEXTEQ,
+				CStringGetTextDatum(stmt->relation->relname));
+	// remove db_name from schema_name for search in bbf_view_def_schemaname
+	curr_schemaname += (strlen(cur_db) + 1);
```
this is wired, if we don't need db name, it should not reach here

macwoody commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2308,3 +2312,278 @@ get_db_owner_role_name(const char *dbname)
+	// search for the row for update => build the key
+	objname_data = (NameData *) palloc0(NAMEDATALEN);
+	snprintf(objname_data->data, NAMEDATALEN, "%s", stmt->subname);
+	ScanKeyInit(&key[0],
```
function name alone is not unique. 
In postgresql's case, nspnanme + function name + funcsignature define a unique function or procedure

macwoody commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2308,3 +2312,278 @@ get_db_owner_role_name(const char *dbname)
+}
+
+static void
+rename_func_update_bbf_catalog(RenameStmt *stmt)
```
why not reuse above function since they share same underlying table.

macwoody commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3012,18 +3012,43 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
+						standard_ProcessUtility(pstmt, queryString, readOnlyTree, context,
+												params, queryEnv, dest, qc);
+					rename_update_bbf_catalog(stmt);
+				}
```
catch and immediate throw is equal to do nothing

macwoody commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3012,18 +3012,43 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
+					else
+						standard_ProcessUtility(pstmt, queryString, readOnlyTree, context,
+												params, queryEnv, dest, qc);
+					rename_update_bbf_catalog(stmt);
```
you could simply leave this function call under TSQL dialect and remove the rest. 
Since the only diff with non-TSQL case is whethere this function is called or not. 
You could even check dialect inside of this function just like check_extra_schema_restrictions

raydhkim committed:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-3591

macwoody commented on test/JDBC/expected/Test-sp_rename-vu-verify.out:
```diff
@@ -0,0 +1,395 @@
+~~END~~
+
+
+EXEC babelfish_sp_rename_internal 'sp_rename_view1_new2', 'sp_rename_view1_new', 'sp_rename_schema1', 'V';
```
we don't have to directly test this

macwoody commented on test/JDBC/expected/Test-sp_rename-vu-verify.out:
```diff
@@ -0,0 +1,395 @@
+~~END~~
+
+
+EXEC sp_rename '', 'sp_rename_view1_new', 'OBJECT';
```
this is an empty object , for your case, you need to fill null value

macwoody commented on test/JDBC/expected/Test-sp_rename-vu-verify.out:
```diff
@@ -0,0 +1,395 @@
+
+
+-- Null input for newname: error, requiring input for newname
+EXEC sp_rename 'sp_rename_view1_new', '', 'OBJECT';
```
same as above

macwoody reviewed (commented):
could you please also add some tests to rename "renamed objects" ?

raydhkim committed:
Revise test case for sp_rename

macwoody commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3012,18 +3012,29 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
-				check_extra_schema_restrictions(parsetree);
+				RenameStmt *stmt = (RenameStmt *) parsetree;
 
+				PG_TRY();
```
this try block is not needed

macwoody commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3012,18 +3012,29 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
+					if (prev_ProcessUtility)
+						prev_ProcessUtility(pstmt, queryString, readOnlyTree, context,
+											params, queryEnv, dest, qc);
+					else
```
at least this function is wrong from my perspective. 

it should be 
				if (prev_ProcessUtility)
					prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
							queryEnv, dest, qc);
				else
					standard_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
							queryEnv, dest, qc);
                check_extra_schema_restrictions			
                 if (dialect = tsql) 
			      rename_update_bbf_catalog
                 return;

raydhkim committed:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-3591

raydhkim committed:
Make vu-test and dep-vu-test more distinguishable

Signed-off-by: Ray Kim <raydhkim@amazon.com>
